lib/puppet/resource/catalog.rb

```diff
@@ -538,7 +538,10 @@ def filter(&block)
 
   # Store the classes in the classfile.
   def write_class_file
-    ::File.open(Puppet[:classfile], "w") do |f|
+    # classfile paths may contain UTF-8
+    # https://docs.puppet.com/puppet/latest/reference/configuration.html#classfile
+    classfile = Puppet.settings.setting(:classfile)
+    Puppet::FileSystem.open(classfile.value, classfile.mode.to_i(8), "w:UTF-8") do |f|
```

Comment:
Note for posterity, decision was to move the application of ignoring mode down a level into `Puppet::FileSystem::Windows`